core/src/main/java/org/infinispan/interceptors/impl/CallInterceptor.java

```diff
@@ -604,16 +608,37 @@ public Object visitRemoveExpiredCommand(InvocationContext ctx, RemoveExpiredComm
 
    @Override
    public Object visitSizeCommand(InvocationContext ctx, SizeCommand command) throws Throwable {
-      ClusterPublisherManager managerToUse;
-      if (command.hasAnyFlag(FlagBitSets.CACHE_MODE_LOCAL)) {
-         managerToUse = localClusterPublisherManager;
-      } else {
-         managerToUse = clusterPublisherManager;
+      CompletionStage<Long> sizeStage = trySizeOptimization(command.getFlagsBitSet())
+            .thenCompose(size -> {
+               if (size > -1) {
+                  return CompletableFuture.completedFuture(size);
+               }
+
+               ClusterPublisherManager<?, ?> managerToUse;
+               if (command.hasAnyFlag(FlagBitSets.CACHE_MODE_LOCAL)) {
+                  managerToUse = localClusterPublisherManager;
+               } else {
+                  managerToUse = clusterPublisherManager;
+               }
+
+               return managerToUse.keyReduction(false, null, null, ctx,
+                     command.getFlagsBitSet(), DeliveryGuarantee.EXACTLY_ONCE, PublisherReducers.count(),
+                     PublisherReducers.add());
+            });
+      return asyncValue(sizeStage);
+   }
+
+   private CompletionStage<Long> trySizeOptimization(long flags) {
```

Comment:
This method can just return a long it doesn't need the CompletionStage. Then in the `visitSizeCommand` check for -1 and if not return the size immediately, otherwise do the code that used to exist in this method.